docs/tactical/DESIGN.md

```diff
@@ -0,0 +1,102 @@
+
+## Constraints and Assumptions
+
+- Tactical is designed to be incrementally integrated — alongside an existing model — into any application, but will ship with additional features for Angular 2 applications.
+- The core library will provide support for both JavaScript and Dart in the browser and on the server.
+- Applications will target browsers with HTML5 storage APIs.
+- Backend services are capable of providing — using such techniques as websockets or polling — an open dialogue with the client.
+- Backend services can support a versioning system capable of validating changes to the model.
+
+## Design
+
+### Data Model
+
+In order to allow end users to be able to interact with their application — even while weakly connected — Tactical provides a local representation of the data model for the application to work with.
+
+To build up this local representation, Tactical has specifications that backend services must meet. In order to make Tactical more adaptable, its specifications only require that backend services be capable of supplying their objects in JSON format. The structure of the JSON is up to each consumer to decide for their respective application. Tactical does not inspect the internal properties of JSON data and will retain any particular order inside the structure.
+
+Once supplied, Tactical will store the received JSON objects in persistent browser storage using JSON objects as keys. Keys are provided by the application when it makes calls to Tactical. Each key should be able to uniquely describe the object that’s associated with it. The representation of a key is up to the application, but can be as simple as a JSON representation of a standard HTTP request body. Tactical will handle translating each key to a unique token that can be used to index the key’s respective value.
+
+However, for most applications, this local representation will not be complete. By registering tactics, consumers will have the ability to easily specify what data would provide a more usable offline data model. From there, Tactical will handle prudently making the appropriate calls in order to eagerly prepare the offline model — reducing application latency in the process.
+
+### Data Manager
+
+Providing this experience requires that Tactical manage three separate representations of the application model: the complete data model _(which is managed by the backend)_, a partial data model _(which is managed by Tactical)_, and changes to the model _(which are managed by the end user)_. Mediating these three representations — in a way that both respects all expectations of the general model and achieves a desirable user experience — will be the most difficult challenge for this library.
+
+As a first step towards solving this challenge, Tactical provides a Data Manager. The Tactical Data Manager will be responsible for:
+
+1. Prudently loading data from the backend in order to build a better local model for the application to use.
```

Comment:
This isn't actually a function of the data manager - prefetching/eager loading is entirely an application/tactic function.